kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncProcessor.java

```diff
@@ -0,0 +1,456 @@
+   * prioritizes moving events through the async processing pipeline over
+   * maximizing the number of events we can get through in each call
+   */
+  private void executeAvailableEvents() {
+    // Start by going through the events waiting to be finalized and finish executing their
+    // outputs, if any, so we can mark them complete and potentially free up blocked events
+    // waiting to be scheduled.
+    drainFinalizingQueue();
+
+    // Then we check the scheduling queue and hand everything that is able to be processed
+    // off to the processing queue for the AsyncThread to continue from here
+    drainSchedulingQueue();
+  }
+
+  private void drainSchedulingQueue() {
+    final List<AsyncEvent> eventsToSchedule = new LinkedList<>();
+
+    while (schedulingQueue.hasProcessableRecord()) {
+      final AsyncEvent processableEvent = schedulingQueue.poll();
+      processableEvent.transitionToInputReady();
+      eventsToSchedule.add(processableEvent);
+    }
+    threadPool.scheduleForProcessing(taskId.partition(), eventsToSchedule);
+  }
+
+  private void drainFinalizingQueue() {
+    while (!finalizableRecords.isEmpty()) {
+      final AsyncEvent event = finalizableRecords.nextFinalizableEvent();
+
+      streamThreadContext.prepareToFinalizeEvent(event.recordContext());
+
+      finalizeEvent(event);
+
+      pendingEvents.remove(event);
+      schedulingQueue.unblockKey(event.inputKey());
+    }
+  }
+
+  private void finalizeEvent(final AsyncEvent event) {
+    event.transitionToFinalizing();
+
+    DelayedWrite<?, ?> nextDelayedWrite = event.nextWrite();
+    DelayedForward<KOut, VOut> nextDelayedForward = event.nextForward();
+    while (nextDelayedWrite != null || nextDelayedForward != null) {
+
+      if (nextDelayedWrite != null) {
+        streamThreadContext.executeDelayedWrite(nextDelayedWrite);
+      }
+
+
+      if (nextDelayedForward != null) {
+        streamThreadContext.executeDelayedForward(nextDelayedForward);
+      }
+
+      nextDelayedWrite = event.nextWrite();
+      nextDelayedForward = event.nextForward();
+    }
+    event.transitionToDone();
+  }
+
+  /**
+   * @return true iff all records have been fully processed from start to finish
+   */
+  private boolean isCompleted() {
+    return pendingEvents.isEmpty();
+  }
+
+  /**
+   * Verify that all the stores accessed by the user via
+   * {@link ProcessorContext#getStateStore(String)} during their processor's
+   * #init method were connected to the processor following the appropriate
+   * procedure for async processors. For more details, see the instructions
+   * in the javadocs for {@link AsyncProcessorSupplier}.
+   */
+  private static void verifyConnectedStateStores(
+      final Map<String, AsyncKeyValueStore<?, ?>> accessedStores,
+      final Map<String, AsyncStoreBuilder<?>> connectedStores,
+      final Logger log
+  ) {
+    if (accessedStores.size() != connectedStores.size()) {
```

Comment:
we could loosen this to just check that everything that was accessed is connected (dont need to check they are exactly the same)

Reply:
Fair enough (to confirm, you meant to keep only this size check but remove the name-based check?)